design/pipeline-artifacts.md

```diff
@@ -47,9 +47,16 @@ Artifacts are downloaded either to `$PIPELINE_RESOURCESDIRECTORY` or to the dire
 
 ```yaml
 - download: string # identifier for the pipeline resource from which to download artifacts, optional; "current" means the current pipeline
-  artifact: string # identifier for the artifact to download; if left blank, downloads all artifacts associated with the resource provided
-  patterns: string | [ string ] # a minimatch path or list of [minimatch paths](tasks/file-matching-patterns.md) to download; if blank, the entire artifact is downloaded
+  artifact: string # identifier for the artifact to download; required
```

Comment:
Without making the `artifact` property required, it introduces confusing rules about how `downloadPath` should work, when multiple artifacts are resolved.

Otherwise, how should it work? And here are the examples that would need to be answered:

For the current pipeline, with multiple artifacts:

```yaml
jobs:
- job: a
  steps:
  - upload: "*.dll"
    artifact: bin
  - upload: "*.pdb"
    artifact: pdb
- job: b
  dependsOn: a
  steps:
  - download: ""
# I would expect this to translate to two tasks:
# - task: …
#    downloadPath: bin
# - task: …
#    downloadPath: pdb
```

For a referenced pipeline:

```yaml
resources:
  pipelines:
  - pipeline: other # let's say it has two artifacts, "bin" and "pdb"
steps:
- download: other
# I would expect this to translate to two tasks:
# - task: …
#    downloadPath: other/bin
# - task: …
#    downloadPath: other/pdb
```

For the current pipeline with multiple artifacts, and a specified `root`:

```yaml
jobs:
- job: a
  steps:
  - upload: "*.dll"
    artifact: bin
  - upload: "*.pdb"
    artifact: pdb
- job: b
  dependsOn: a
  steps:
  - download: ""
    root: asdf
# I would expect this to translate to two tasks:
# - task: …
#    downloadPath: asdf/bin              # IS THIS RIGHT?
# - task: …
#    downloadPath: asdf/pdb
```

For a referenced pipeline, and a specified `root`:

```yaml
resources:
  pipelines:
  - pipeline: other # let's say it has two artifacts, "bin" and "pdb"
steps:
- download: other
   root: asdf
# I would expect this to translate to two tasks:
# - task: …
#    downloadPath: asdf/bin            # IS THIS RIGHT?
# - task: …
#    downloadPath: asdf/pdb
```

For the current pipeline, a specific artifact, and a specified `root`:

```yaml
jobs:
- job: a
  steps:
  - upload: "*.dll"
    artifact: bin
  - upload: "*.pdb"
    artifact: pdb
- job: b
  dependsOn: a
  steps:
  - download: ""
    artifact: bin
    root: asdf
# I would expect this to translate to one task:
# - task: …
#    downloadPath: asdf              # IS THIS RIGHT?
```

Reply:
(I don't have a strong opinion here, other than we should be predictable and flexible.)

I was nodding along to all your examples until the last one. I _think_ that should translate to
```yaml
- task: ...
  inputs:
    downloadPath: asdf/bin
```

@ashokirla, thoughts?

Reply:
My point here, is actually that the artifact property should be `required`. Then you eliminate the confusion I described.

Reply:
I think it is more of a convenience rather than confusion. 

If we look at the scenarios. 
Scenario 1: User would like to download all the artifacts from `pipeline` resource to a specific destination folder.
Scenario 2: User would like to download files with a given pattern 'SmarHotel**.tar.gz' across all the artifacts from the `pipeline` resource.

These are very valid scenarios. And by making the artifact input optional,
1. User doesn't have to define one section each for each of the artifact.
2. User doesn't have to be aware of the exact name of the artifact published. 
Here is one of the example RD today where user provides the file input from artifact without even knowing the name of the artifact.
![image](https://user-images.githubusercontent.com/25578249/50690526-a64e7380-1053-11e9-9c75-08efe64a229d.png)

Infact there is a feedback from the team to provide a default syntax that can work for both current pipeline and other pipeline resource together

i.e.
```yaml
download: current
root: blah
# This downloads artifacts from current pipeline to a specific destination.
```
```yaml
download: other
root: blah
# This downloads artifacts from 'other' pipeline to a specific destination
```
```yaml
download:
root: blah
#This download artifacts from both other and current pipeline(all the available azure pipelines).
```
We are now using the macro `download` only for azure pipelines. And we also decided to use the keyword 'current' for current pipeline. So we can reserve "" (empty) for referring both internal and external pipeline resources.

Reply:
How common is the requirement to download a set of files across all artifacts? It seems like a corner case and how do you deal with collisions?

Reply:
I merged this PR, but I don't think `artifact` should be required. I'll put some examples in the PR I'm working on now.